### DISCUSSION TITLE
Spark NLP 5.0.0: Introducing ONNX Support, State-of-the-Art Instructor Embeddings, E5 Text Embeddings, Document Similarity Ranker, and Much More!

### DISCUSSION THREAD

**Discussion by maziyarpanahi** [JohnSnowLabs/spark-nlp#13877] [Announcement]
# :loudspeaker: It's All About That Search!

We are delighted to announce the release of **Spark NLP 🚀 5.0.0**, featuring the highly anticipated support for **ONNX**! From the start of 2023, we have been working tirelessly to ensure that the integration of ONNX is not just possible but also seamless for all our users. With this support, you can look forward to faster inference, automatic optimization, and quantization when exporting your LLM models. Additionally, we are also set to release an array of new LLM models fine-tuned specifically for **chat** and **instruction**, now that we have successfully integrated ONNX Runtime into Spark NLP 🚀.

We have introduced two state-of-the-art models for text embedding, `INSTRUCTOR` and `E5` embeddings. Currently, these models are leading the way on the MTEB leaderboard, even outperforming the widely recognized OpenAI [text-embedding-ada-002](https://beta.openai.com/docs/guides/embeddings/types-of-embedding-models). These cutting-edge models are now being utilized in production environments to populate Vector Databases. In addition, they are being paired with LLM models like Falcon, serving to augment their existing knowledge base and reduce the chances of hallucinations.

We want to thank our community for their valuable feedback, feature requests, and contributions. Our [Models Hub](https://nlp.johnsnowlabs.com/models?edition=Spark+NLP) now contains over 18,000+ free and truly open-source models & pipelines. 🎉

----------------
:fire: New Features
----------------

## Spark NLP :heart: ONNX
![SPARK NLP](https://github.com/JohnSnowLabs/spark-nlp/assets/5762953/9517a5ee-ee10-481c-b1db-503662b56308)

**NEW:** Introducing support for **ONNX Runtime** in **Spark NLP**🚀. Serving as a high-performance inference engine, **ONNX Runtime** can handle machine learning models in the ONNX format and has been proven to significantly boost inference performance across a multitude of models.

Our integration of ONNX Runtime has already led to substantial improvements when serving our LLM models, including BERT. Furthermore, this integration empowers Spark NLP users to optimize their model performance. As users export their models to ONNX, they can utilize the built-in features provided by libraries such as `onnx-runtime`, `transformers`, `optimum`, and `PyTorch`. Notably, these libraries offer out-of-the-box capabilities for optimization and quantization, enhancing model efficiency and performance.

<img width="967" alt="image" src="https://github.com/JohnSnowLabs/spark-nlp/assets/5762953/acb4d4ff-5b76-4f89-854b-789200b2ca04">

In the realm of Vector Databases, the quest for faster and more efficient Embeddings models has become an imperative pursuit. Models like BERT, DistilBERT, and DeBERTa have revolutionized natural language processing tasks by capturing intricate semantic relationships between words. However, their computational demands and slow inference times pose significant challenges in the game of Vector Databases.

In Vector Databases, the speed at which queries are processed and embeddings are retrieved directly impacts the overall performance and responsiveness of the system. As these databases store vast amounts of vectorized data, such as documents, sentences, or entities, swiftly retrieving relevant embeddings becomes paramount. It enables real-time applications like search engines, recommendation systems, sentiment analysis, and chat/instruct-like products similar to ChatGPT to deliver timely and accurate results, ensuring a seamless user experience.

Keeping this in mind, we've initiated ONNX support for the following annotators:

- We've introduced ONNX support for the `BertEmbeddings` annotator. Approximately 180 models of the same name have already been converted to the ONNX format to automatically benefit from the associated performance enhancements.
- We've added ONNX support for the `RoBertaEmbeddings` annotator. Roughly 55 models of the same name have been imported in the ONNX format, thus allowing for automatic speed improvements.
- ONNX support has been initiated for the `DistilBertEmbeddings` annotator. Around 25 models with the same name have been converted to the ONNX format, facilitating automatic speed enhancements.
- We've incorporated ONNX support into the `DeBertaEmbeddings` annotator. About 12 models bearing the same name have been imported in the ONNX format, enabling them to automatically reap the benefits of speed improvements.

We have successfully identified all existing models for these annotators on our Models Hub, converted them into the ONNX format, and re-uploaded them under the same names. This process was carried out to ensure a seamless transition for our community starting with Spark NLP 5.0.0. We will continue to import additional models in the ONNX format in the days ahead. To keep track of the ONNX compatibility with Spark NLP, follow this issue: https://github.com/JohnSnowLabs/spark-nlp/issues/13866.

## INSTRUCTOR: Instruction-Finetuned Text Embeddings

**NEW:** Introducing **InstructorEmbeddings** annotator in Spark NLP 🚀. `InstructorEmbeddings` can load new state-of-the-art INSTRUCTOR Models inherited from `Google T5` for Text embedding.

This annotator is compatible with all the models trained/fine-tuned by using `T5EncoderModel` for **PyTorch** or `TFT5EncoderModel` for **TensorFlow** models in HuggingFace 🤗

<img width="1897" alt="image" src="https://github.com/JohnSnowLabs/spark-nlp/assets/5762953/0957f103-fb91-441e-af7a-66f6575f7ac4">

> **Instructor**👨‍🏫, an instruction-finetuned text embedding model that can generate text embeddings tailored to any task (e.g., classification, retrieval, clustering, text evaluation, etc.) and domains (e.g., science, finance, etc.) ***by simply providing the task instruction, without any finetuning***. Instructor👨‍ achieves sota on 70 diverse embedding tasks!
For more details, check out the [official paper](https://arxiv.org/abs/2212.09741) and the [project page](https://instructor-embedding.github.io/)! 

**NOTEBOOK:**  https://github.com/JohnSnowLabs/spark-nlp/blob/master/examples/python/annotation/text/english/sentence-embeddings/InstructorEmbeddings.ipynb

## E5: Text Embeddings by Weakly-Supervised Contrastive Pre-training

**NEW:** Introducing **E5Embeddings** annotator in Spark NLP 🚀. `E5Embeddings` can load new state-of-the-art E5 Models based on BERT for Text Embeddings.

<img width="1187" alt="image" src="https://github.com/JohnSnowLabs/spark-nlp/assets/5762953/e8907ec1-d0e5-4b97-8fe6-84101a295689">

>[Text Embeddings by Weakly-Supervised Contrastive Pre-training](https://arxiv.org/pdf/2212.03533.pdf).
Liang Wang, Nan Yang, Xiaolong Huang, Binxing Jiao, Linjun Yang, Daxin Jiang, Rangan Majumder, Furu Wei, arXiv 2022

**NOTEBOOK:** https://github.com/JohnSnowLabs/spark-nlp/blob/master/examples/python/annotation/text/english/sentence-embeddings/E5Embeddings.ipynb

Our new state-of-the-art annotators for Text Embeddings are currently dominating the top of the MTEB leaderboard positioning themselves above OpenAI [text-embedding-ada-002](https://beta.openai.com/docs/guides/embeddings/types-of-embedding-models)
<img width="1503" alt="image" src="https://github.com/JohnSnowLabs/spark-nlp/assets/5762953/50a5ca09-6867-4d3b-b405-3794ef8d1183">

>Massive Text Embedding Benchmark (MTEB) Leaderboard. To submit, refer to the [MTEB GitHub repository](https://github.com/embeddings-benchmark/mteb#leaderboard) 🤗

## Document Similarity Ranker by LSH techniques
**NEW:** Introducing **DocumentSimilarityRanker** annotator in Spark NLP 🚀. `DocumentSimilarityRanker` is a new annotator that uses LSH techniques present in Spark ML lib to execute approximate nearest neighbors search on top of sentence embeddings, It aims to capture the semantic meaning of a document in a dense, continuous vector space and return it to the ranker search.

**NOTEBOOK:** https://github.com/JohnSnowLabs/spark-nlp/blob/master/examples/python/annotation/text/english/text-similarity/doc-sim-ranker/test_doc_sim_ranker.ipynb

* Welcoming 6 new **Databricks** runtimes to our Spark NLP family:
  * Databricks 13.1 LTS
  * Databricks 13.1 LTS ML
  * Databricks 13.1 LTS ML GPU  
  * Databricks 13.2 LTS
  * Databricks 13.2 LTS ML
  * Databricks 13.2 LTS ML GPU  
* Welcome **AWS EMR** 6.11 version to our Spark NLP family
* Fix BART issue with input longer than the `maxInputLength` 

:floppy_disk: Models
--------------------
Spark NLP 5.0.0 comes with more than **400+** new Large Language Models (LLMs) in ONNX format. We are also providing `optimized` and `quantized` versions of popular models that can be used immediately in any Spark NLP pipelines:

### Featured Models

| Model                | Name               | Lang |  
|:---------------------|:-------------------|:---|
| BertEmbeddings | [bert_base_cased](https://sparknlp.org/2023/06/29/bert_base_cased_en.html) |      `en`
| BertEmbeddings | [bert_base_cased_opt](https://sparknlp.org/2023/06/29/bert_base_cased_opt_en.html) |      `en`
| BertEmbeddings | [bert_base_cased_quantized](https://sparknlp.org/2023/06/29/bert_base_cased_quantized_en.html) |      `en`
| BertEmbeddings | [small_bert_L2_768](https://sparknlp.org/2023/06/28/small_bert_L2_768_en.html) |      `en`
| BertEmbeddings | [small_bert_L2_768_opt](https://sparknlp.org/2023/06/28/small_bert_L2_768_opt_en.html) |      `en`
| BertEmbeddings | [small_bert_L2_768_quantized](https://sparknlp.org/2023/06/28/small_bert_L2_768_quantized_en.html) |      `en`
| DeBertaEmbeddings | [roberta_base](https://sparknlp.org/2023/06/28/roberta_base_en.html) |      `en`
| DeBertaEmbeddings | [roberta_base_opt](https://sparknlp.org/2023/06/28/roberta_base_opt_en.html) |      `en`
| DeBertaEmbeddings | [roberta_base_quantized](https://sparknlp.org/2023/06/28/roberta_base_quantized_en.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_cased](https://sparknlp.org/2023/06/28/distilbert_base_cased_en.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_cased_opt](https://sparknlp.org/2023/06/28/distilbert_base_cased_opt_en.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_cased_quantized](https://sparknlp.org/2023/06/28/distilbert_base_cased_quantized_en.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_uncased](https://sparknlp.org/2023/06/28/distilbert_base_uncased_en.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_uncased_opt](https://sparknlp.org/2023/06/28/distilbert_base_uncased_opt_en.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_uncased_quantized](https://sparknlp.org/2023/06/28/distilbert_base_uncased_quantized_en.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_multilingual_cased](https://sparknlp.org/2023/06/28/distilbert_base_multilingual_cased_xx.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_multilingual_cased_opt](https://sparknlp.org/2023/06/28/distilbert_base_multilingual_cased_opt_xx.html) |      `en`
| DistilBertEmbeddings | [distilbert_base_multilingual_cased_quantized](https://sparknlp.org/2023/06/28/distilbert_base_multilingual_cased_quantized_xx.html) |      `en`
| RoBertaBertEmbeddings | [roberta_base](https://sparknlp.org/2023/06/28/roberta_base_en.html) |      `en`
| RoBertaBertEmbeddings | [roberta_base_opt](https://sparknlp.org/2023/06/28/roberta_base_opt_en.html) |      `en`
| RoBertaBertEmbeddings | [roberta_base_quantized](https://sparknlp.org/2023/06/28/roberta_base_quantized_en.html) |      `en`
| InstructorEmbeddings | [instructor_base](https://sparknlp.org/2023/06/08/instructor_base_en.html) |      `en`
| InstructorEmbeddings | [instructor_large](https://sparknlp.org/2023/06/21/instructor_large_en.html) |      `en`
| E5Embeddings | [e5_small](https://sparknlp.org/2023/06/21/e5_small_en.html) |      `en`
| E5Embeddings | [e5_small_v2](https://sparknlp.org/2023/06/21/e5_small_v2_en.html) |      `en`
| E5Embeddings | [e5_base](https://sparknlp.org/2023/06/21/e5_base_en.html) |      `en`
| E5Embeddings | [e5_base_v2](https://sparknlp.org/2023/06/21/e5_base_v2_en.html) |      `en`
| E5Embeddings | [e5_large](https://sparknlp.org/2023/06/21/e5_large_en.html) |      `en`
| E5Embeddings | [e5_large_v2](https://sparknlp.org/2023/06/21/e5_large_v2_en.html) |      `en`

### :notebook: New Notebooks

| Notebooks | Colab |
| :-----------| :------|
|Instructor Embeddings|[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JohnSnowLabs/spark-nlp/blob/master/examples/python/annotation/text/english/sentence-embeddings/InstructorEmbeddings.ipynb)
|E5 Embeddings|[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JohnSnowLabs/spark-nlp/blob/master/examples/python/annotation/text/english/sentence-embeddings/E5Embeddings.ipynb)
|Document Similarity Ranker|[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/JohnSnowLabs/spark-nlp/blob/master/examples/python/annotation/text/english/text-similarity/doc-sim-ranker/test_doc_sim_ranker.ipynb)

- You can visit [Import Transformers in Spark NLP](https://github.com/JohnSnowLabs/spark-nlp/discussions/5669)
- You can visit [Spark NLP Examples](https://github.com/JohnSnowLabs/spark-nlp/tree/master/examples) for 100+ examples

----------------
:book: Documentation
----------------
* [Import models from TF Hub & HuggingFace](https://github.com/JohnSnowLabs/spark-nlp/discussions/5669) 
* [Spark NLP Notebooks](https://github.com/JohnSnowLabs/spark-nlp/tree/master/examples)
* [Models Hub](https://nlp.johnsnowlabs.com/models?edition=Spark+NLP) with new models
* [Spark NLP Articles](https://medium.com/spark-nlp)
* [Spark NLP in Action](https://nlp.johnsnowlabs.com/demo)
* [Spark NLP Documentation](https://nlp.johnsnowlabs.com/docs/en/quickstart)
* [Spark NLP Scala APIs](https://nlp.johnsnowlabs.com/api)
* [Spark NLP Python APIs](https://nlp.johnsnowlabs.com/api/python)

----------------
❤️  Community support
----------------

- [Slack](https://join.slack.com/t/spark-nlp/shared_invite/zt-198dipu77-L3UWNe_AJ8xqDk0ivmih5Q) For live discussion with the Spark NLP community and the team
- [GitHub](https://github.com/JohnSnowLabs/spark-nlp) Bug reports, feature requests, and contributions
- [Discussions](https://github.com/JohnSnowLabs/spark-nlp/discussions) Engage with other community members, share ideas,
  and show off how you use Spark NLP!
- [Medium](https://medium.com/spark-nlp) Spark NLP articles
- JohnSnowLabs official [Medium](https://medium.com/john-snow-labs)
- [YouTube](https://www.youtube.com/channel/UCmFOjlpYEhxf_wJUDuz6xxQ/videos) Spark NLP video tutorials

---------------
Installation
---------------
**Python**

```shell
#PyPI

pip install spark-nlp==5.0.0
```

**Spark Packages**

**spark-nlp** on Apache Spark 3.0.x, 3.1.x, 3.2.x, 3.3.x, and 3.4.x (Scala 2.12):

```shell
spark-shell --packages com.johnsnowlabs.nlp:spark-nlp_2.12:5.0.0

pyspark --packages com.johnsnowlabs.nlp:spark-nlp_2.12:5.0.0
```

**GPU**

```shell
spark-shell --packages com.johnsnowlabs.nlp:spark-nlp-gpu_2.12:5.0.0

pyspark --packages com.johnsnowlabs.nlp:spark-nlp-gpu_2.12:5.0.0
```

**Apple Silicon (M1 & M2)**
```shell
spark-shell --packages com.johnsnowlabs.nlp:spark-nlp-silicon_2.12:5.0.0

pyspark --packages com.johnsnowlabs.nlp:spark-nlp-silicon_2.12:5.0.0
```

**AArch64**
```shell
spark-shell --packages com.johnsnowlabs.nlp:spark-nlp-aarch64_2.12:5.0.0

pyspark --packages com.johnsnowlabs.nlp:spark-nlp-aarch64_2.12:5.0.0
```

**Maven**

**spark-nlp** on Apache Spark 3.0.x, 3.1.x, 3.2.x, 3.3.x, and 3.4.x:

```xml
<dependency>
    <groupId>com.johnsnowlabs.nlp</groupId>
    <artifactId>spark-nlp_2.12</artifactId>
    <version>5.0.0</version>
</dependency>
```

**spark-nlp-gpu:**

```xml
<dependency>
    <groupId>com.johnsnowlabs.nlp</groupId>
    <artifactId>spark-nlp-gpu_2.12</artifactId>
    <version>5.0.0</version>
</dependency>
```

**spark-nlp-silicon:**

```xml
<dependency>
    <groupId>com.johnsnowlabs.nlp</groupId>
    <artifactId>spark-nlp-silicon_2.12</artifactId>
    <version>5.0.0</version>
</dependency>
```

**spark-nlp-aarch64:**

```xml
<dependency>
    <groupId>com.johnsnowlabs.nlp</groupId>
    <artifactId>spark-nlp-aarch64_2.12</artifactId>
    <version>5.0.0</version>
</dependency>
```

**FAT JARs**

* CPU on Apache Spark 3.x/3.1.x/3.2.x/3.3.x/3.4.x: https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/jars/spark-nlp-assembly-5.0.0.jar

* GPU on Apache Spark 3.0.x/3.1.x/3.2.x/3.3.x/3.4.x: https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/jars/spark-nlp-gpu-assembly-5.0.0.jar

* M1 on Apache Spark 3.0.x/3.1.x/3.2.x/3.3.x/3.4.x: https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/jars/spark-nlp-silicon-assembly-5.0.0.jar

* AArch64 on Apache Spark 3.0.x/3.1.x/3.2.x/3.3.x/3.4.x: https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/jars/spark-nlp-aarch64-assembly-5.0.0.jar

## What's Changed
* SPARKNLP 836 - Introducing "Instructor Embeddings" for sentence embeddings like Instructor-XL model by @prabod in https://github.com/JohnSnowLabs/spark-nlp/pull/13849
* Integrating ONNX runtime (ORT) in Spark NLP 5.0.0 🎉 by @maziyarpanahi in https://github.com/JohnSnowLabs/spark-nlp/pull/13857
* SPARKNLP 852 - Introducing "E5 Embeddings" for sentence embeddings like e5-large-v2 model by @prabod in https://github.com/JohnSnowLabs/spark-nlp/pull/13859
* SPARKNLP-846: BART: Added maxInputLength. by @prabod in https://github.com/JohnSnowLabs/spark-nlp/pull/13863
* Feature/doc similarity ranker by @wolliq in https://github.com/JohnSnowLabs/spark-nlp/pull/13858
* Added Sentence Embeddings Notebooks. by @prabod in https://github.com/JohnSnowLabs/spark-nlp/pull/13874
* Chore: conda recipe update by @luca-martial in https://github.com/JohnSnowLabs/spark-nlp/pull/13764
* Models hub by @maziyarpanahi @ahmedlone127 in https://github.com/JohnSnowLabs/spark-nlp/pull/13876
* Release/500 release candidate by @maziyarpanahi in https://github.com/JohnSnowLabs/spark-nlp/pull/13873

**Full Changelog**: https://github.com/JohnSnowLabs/spark-nlp/compare/4.4.4...5.0.0

**Comment by AayushSameerShah** [comment]
Lovved it!!
Thank you so much for releasing this anticipated feature! 
🙌🏻

**Reply by maziyarpanahi** [reply]
Now on to the CasualLM like `GPT-NeoX` 🚀!

**Comment by nguyenvanlinh1808** [comment]
I can not download [spark-nlp-gpu-assembly-5.0.0.jar](https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/jars/spark-nlp-gpu-assembly-5.0.0.jar), if you have a review on that, thanks

**Reply by maziyarpanahi** [reply]
Thanks @nguyenvanlinh1808 
The upload script failed to put those on S3, I'll re-launch them today

**Reply by maziyarpanahi** [reply]
It has been uploaded now, thanks for reporting it 

[spark-nlp-gpu-assembly-5.0.0.jar](https://s3.amazonaws.com/auxdata.johnsnowlabs.com/public/jars/spark-nlp-gpu-assembly-5.0.0.jar)
